src/Symfony/Component/Filesystem/Watcher/FileChangeWatcher.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Filesystem\Watcher;
+
+use Symfony\Component\Filesystem\Watcher\Resource\Locator\FileResourceLocator;
+
+/**
+ * @author Pierre du Plessis <pdples@gmail.com>
+ *
+ * @internal
+ */
+final class FileChangeWatcher implements WatcherInterface
+{
+    private $locator;
+
+    public function __construct()
+    {
+        $this->locator = new FileResourceLocator();
+    }
+
+    public function watch($path, callable $callback, float $timeout = null): void
+    {
+        $resource = $this->locator->locate($path);
+
+        if (!$resource) {
+            throw new \InvalidArgumentException(sprintf('"%s" is not a valid path to watch.', \gettype($path)));
+        }
+
+        $run = true;
+        $start = microtime(true);
+
+        while ($run) {
+            if ($changes = $resource->detectChanges()) {
+                foreach ($changes as $change) {
+                    $run = false !== $callback($change->getFile(), $change->getEvent());
+                }
+
+                $start = microtime(true);
+            }
+
+            if (null !== $timeout && ($timeout / 1000) <= (microtime(true) - $start)) {
```

Comment:
Having configurable timeout here won't be beneficial, as this class is internal and should not be used by any user. Users will just use the `Filesystem::watch` method, which has an argument for a timeout after which the watch should stop if no events happening. There is also a sleep below, so it won't just loop over and over without waiting

Reply:
Ah I see where is misunderstanding there. You implemented timeout as in, time after which FileChangeWatcher should be aborted. I don't see what purpose this would have. This is also not something that other, inotify implementation does. Request was to implement configurable time for which loop blocks, not time after which loop is terminated. This is important to make watcher pick up changes faster when desired, right now it will take about 1 second.

Reply:
This was a suggestion from @nicolas-grekas to break the loop after the timeout (https://github.com/symfony/symfony/pull/31462#discussion_r342699297).

Making the sleep configurable will only be beneficial if `inotify` is not available, so I don't want to add another generic parameter just for one implementation

Reply:
If you don't want to add another generic parameter to FileSystem class, we need to unmark internal of this class and make it configurable via constructor. Configurable sleep is important.